### PR TITLE
Change Bound to &Bound

### PR DESCRIPTION
**Pull request by Stargateur**
POC about can we keep a `&Bound` and not a `Bound` to deserialize.

### CODE REVIEW

File: src/de.rs [hunk 1]
```diff
@@ -1,24 +1,11 @@
...
-where
-    T: Deserialize<'de>,
-{
-    let mut depythonizer = Depythonizer::from_object_bound(obj.as_borrowed().to_owned());
-    T::deserialize(&mut depythonizer)
-}
-
-/// Attempt to convert a Python object to an instance of `T`
-pub fn depythonize_bound<'py, T>(obj: Bound<'py, PyAny>) -> Result<T>
+pub fn depythonize_bound<'py, T>(obj: &Bound<'py, PyAny>) -> Result<T>
```
Comment (davidhewitt) [inline src/de.rs:8]: I'm happy to remove the original `depythonize` in the next release. Maybe instead can we please add `depythonize` back with this new `&Bound` as input, and mark `depythonize_bound` as deprecated?

The goal with the PyO3 0.21 GIL Refs removal is that in 0.23 when we've finished the migration we will rename `PyTuple::new_bound` to just `PyTuple::new` etc. in the same way.

File: src/de.rs [hunk 2]
```diff
@@ -27,34 +14,28 @@ where
...
-        note = "will be replaced by `Depythonizer::from_object_bound` in a future version"
-    )]
-    pub fn from_object(input: &'py PyAny) -> Self {
-        Self::from_object_bound(input.as_borrowed().to_owned())
-    }
-
+impl<'py, 'bound> Depythonizer<'py, 'bound> {
     /// Create a deserializer from a Python object
-    pub fn from_object_bound(input: Bound<'py, PyAny>) -> Self {
+    pub fn from_object_bound(input: &'bound Bound<'py, PyAny>) -> Self {
```
Comment (davidhewitt) [inline src/de.rs:23]: Similarly let's deprecate `from_object_bound` and replace `from_object` instead.

**Review by davidhewitt — COMMENTED** [review]
Thanks, this is a good change in my opinion!

File: src/de.rs (hunk 2 above)
Comment (Stargateur) [reply src/de.rs:23]: I can't implement for `Bound` the item would be dropped. Anyway, just add a lifetime parameter to `Depythonizer` struct is breaking change. What we can do is have two `Depythonizer` struct one with the ref the other without.

**Review by davidhewitt — APPROVED** [review]
Thanks, sorry for the long delay on my part here.

**Pull request merged**